Introduce preliminary macro operation fusion

qwe661234 opened the pull request:
Through our observations, we have identified certain patterns in instruction sequences. By converting these specific RISC-V instruction patterns into faster and equivalent code, we can significantly improve execution efficiency.

In our current analysis, we focus on a commonly used benchmark and have found the following frequently occurring instruction patterns: auipc + addi, auipc + add, multiple sw, and multiple lw.

|Metric| commit fba5802 | macro fuse operation |Speedup|
|--------| -------- | -------- | -------- | 
| CoreMark | 1351.065 (Iterations/Sec)| 1352.843 (Iterations/Sec) |+0.13％|
| dhrystone  |1073 DMIPS |1146 DMIPS |+6.8%|
| nqueens |8295 msec    | 7824  msec   |+6.0%|

jserv commented on src/decode.h:
```diff
@@ -156,7 +156,12 @@
         _(cswsp, 0)                        \
-    )
+    )                                      \
+    _(fuse1, 0)                            \
```
Add a comment,
> macro operation fusion: convert specific RISC-V instruction patterns into shorter and equivalent code

jserv replied:
In JVM, such technique is called "quickening optimization."
See https://www.complang.tuwien.ac.at/andi/papers/dotnet_06.pdf

jserv commented on src/decode.h:
```diff
@@ -228,6 +233,11 @@ enum {
+typedef struct mem_fuse {
+    int32_t imm;
+    uint8_t rd, rs1, rs2;
+} mem_fuse_t;
```
Name it `opcode_fuse_t`.

jserv commented on src/decode.h:
```diff
@@ -228,6 +233,11 @@ enum {
     INSN_32 = 4,
 };
 
+typedef struct mem_fuse {
```
Drop "mem_fuse" since it is not necessary.

jserv commented on src/emulate.c:
```diff
@@ -1407,6 +1449,92 @@ static void extend_block(riscv_t *rv, block_t *block)
+                    break;
+                count++;
+            }
+            if (count >= 5) {
```
Explain why 5 appears here.

jserv commented on src/emulate.c:
```diff
@@ -1407,6 +1449,92 @@ static void extend_block(riscv_t *rv, block_t *block)
+            }
+            if (count >= 5) {
+                ir->opcode = rv_insn_fuse3;
+                ir->mem_fuse = malloc(count * sizeof(mem_fuse_t));
```
Utilize memory pool when possible.

jserv replied:
When can we release the memory occupied by opcode fusion?

qwe661234 replied:
When we free a basic block with opcode_fuse.

jserv commented on src/emulate.c:
```diff
@@ -1407,6 +1449,92 @@ static void extend_block(riscv_t *rv, block_t *block)
         last_ir->branch_untaken = next->ir;
 }
 
+static void match_pattern(block_t *block)
```
Add some comments.

jserv commented on src/emulate.c:
```diff
@@ -1407,6 +1449,92 @@ static void extend_block(riscv_t *rv, block_t *block)
+                }
+            }
+            break;
+        case rv_insn_lw:
```
Avoid code duplication for `lw` and `sw`.

jserv commented:
> To enhance execution efficiency, we employ instruction fusion by combining sequences that adhere to specific patterns into fused instructions. Currently, we have incorporated four fused instructions: auipc + addi, auipc + add, multiple sw, and multiple lw.

You shall show some numbers to illustrate how we can benefit from macro operation fusion.
In addition, why were 4 patterns were picked? Denote them with existing benchmark programs.

jserv commented on src/emulate.c:
```diff
@@ -1277,6 +1277,48 @@ RVOP(cswsp, {
+    mem_fuse_t *mem_fuse = ir->mem_fuse;
+    for (int i = 0; i < ir->imm2; i++) {
+        const uint32_t addr = rv->X[mem_fuse[i].rs1] + mem_fuse[i].imm;
+        RV_EXC_MISALIGN_HANDLER(3, store, false, 1);
```
This misalign handler might hurt performance. Can you validate memory region in advance? That is, write memory contiguously when possible.

qwe661234 replied:
We only check misaligned for the first load or store instruction.

jserv commented on src/emulate.c:
```diff
@@ -1277,6 +1277,48 @@ RVOP(cswsp, {
+    mem_fuse_t *mem_fuse = ir->mem_fuse;
+    for (int i = 0; i < ir->imm2; i++) {
+        const uint32_t addr = rv->X[mem_fuse[i].rs1] + mem_fuse[i].imm;
+        RV_EXC_MISALIGN_HANDLER(3, load, false, 1);
```
Ditto. Think of the possibility to fetch memory contiguously.

jserv commented on src/emulate.c:
```diff
@@ -1407,6 +1449,92 @@ static void extend_block(riscv_t *rv, block_t *block)
+                ir->impl = dispatch_table[ir->opcode];
+                for (int j = 1; j < count; j++) {
+                    next_ir = ir + j;
+                    ir->mem_fuse[j].imm = next_ir->imm;
```
Instead of various assignments, can you utilize `memcpy` for structure member filling? `memcpy` usually outperforms.

jserv commented on src/decode.h:
```diff
@@ -240,6 +253,9 @@ typedef struct rv_insn {
 #endif
+    /* fuse operation */
+    int32_t imm2;
+    opcode_fuse_t *opcode_fuse;
```
Rename the variable name to `fuse` for simplicity.

jserv commented on src/riscv.c:
```diff
@@ -25,6 +25,10 @@ void block_map_clear(block_map_t *map)
         if (!block)
             continue;
+        for (uint32_t i = 0; i < block->n_insn; i++) {
+            if (block->ir[i].fuse)
```
Simply remove `if (block->ir[i].fuse)` since [free(3p)](https://man7.org/linux/man-pages/man3/free.3p.html) would check.
> If ptr is a null pointer, no action shall occur.

jserv commented on src/emulate.c:
```diff
@@ -1356,6 +1410,78 @@ static void block_translate(riscv_t *rv, block_t *block)
     block->ir[block->n_insn - 1].tailcall = true;
 }
 
+#define pack_memory_operation(RW)                                          \
```
Capitalize macro name. That is, `PACK_MEM_OPS`.

jserv replied:
The verb "pack" was not precise. Instead, use "combine," which conveys the notion of merging multiple instructions into a single one.

jserv commented on src/emulate.c:
```diff
@@ -1356,6 +1410,78 @@ static void block_translate(riscv_t *rv, block_t *block)
+             * data are contiguous, pack these instructions. */
+            pack_memory_operation(0);
+        case rv_insn_lw:
+            /* If the memory addresses of a sequence of load instructions for
```
Remove this comment. Write comments prior to `case rv_insn_sw` and `case rv_insn_lw`.

jserv commented on src/decode.h:
```diff
@@ -240,6 +253,9 @@ typedef struct rv_insn {
     uint8_t shamt;
 #endif
+    /* fuse operation */
+    int32_t imm2;
```
Can we use narrow data types for `imm2`?

jserv commented on src/emulate.c:
```diff
@@ -1219,6 +1220,60 @@ RVOP(cswsp, {
+    rv->PC += ir->insn_len * (ir->imm2 - 1);
+})
+
+static bool do_empty(riscv_t *rv, const rv_insn_t *ir)
```
Can we unify `empty` and `nop`?

jserv commented on src/emulate.c:
```diff
@@ -1219,6 +1220,60 @@ RVOP(cswsp, {
 })
 #endif
 
+/* auipc + addi */
```
Is it possible to manipulate the sequence `lui` + `addi`?
See https://github.com/sysprog21/rv32emu/issues/81#issuecomment-1318209733

Disassembly of CoreMark:
```
   10324:       000087b7                lui     a5,0x8
   10328:       b0578793                addi    a5,a5,-1275 # 0x7b05
```

qwe661234 replied:
It is possible. however, there are some problems when running qrcode.elf if we import this pattern, so I skip it in this pull request.

jserv replied:
> It is possible. however, there are some problems when running qrcode.elf if we import this pattern, so I skip it in this pull request.

Add a comment starting with "FIXME: lui + addi"

jserv commented on src/emulate.c:
```diff
@@ -1219,6 +1220,60 @@ RVOP(cswsp, {
+    rv->PC += ir->insn_len * (ir->imm2 - 1);
+})
+
+/* multiple lw */
```
`lw` is the most frequent instruction (see #34), and we might dive into its use case more.

jserv replied:
Can you handle the following case? (disassembly from CoreMark
```
   10248:       03012603                lw      a2,48(sp)
   1024c:       01c11583                lh      a1,28(sp)
   10250:       03412503                lw      a0,52(sp)
```

jserv replied:
In addition, consider the following scenario:
```
   10a84:       01c12083                lw      ra,28(sp)
   10a88:       07f47513                andi    a0,s0,127
   10a8c:       01812403                lw      s0,24(sp)
   10a90:       01412483                lw      s1,20(sp)
   10a94:       01012903                lw      s2,16(sp)
   10a98:       00c12983                lw      s3,12(sp)
```

It can be regarded as 5 `lw`. Roughly speaking, if [peephole optimization](https://en.wikipedia.org/wiki/Peephole_optimization) can be applied, we shall benefit from further optimizations.

jserv replied:
Another case: (disassembly from CoreMark)
```
   10c08:       01162023                sw      a7,0(a2)
   10c0c:       00052783                lw      a5,0(a0)
   10c10:       00059883                lh      a7,0(a1)
   10c14:       00259603                lh      a2,2(a1)
   10c18:       00f82023                sw      a5,0(a6)
   10c1c:       01052023                sw      a6,0(a0)
   10c20:       00e82223                sw      a4,4(a6)
   10c24:       0006a783                lw      a5,0(a3)
```

Mixture of `sw` and `lw`.

qwe661234 replied:
> Another case: (disassembly from CoreMark)
> 
> ```
>    10c08:       01162023                sw      a7,0(a2)
>    10c0c:       00052783                lw      a5,0(a0)
>    10c10:       00059883                lh      a7,0(a1)
>    10c14:       00259603                lh      a2,2(a1)
>    10c18:       00f82023                sw      a5,0(a6)
>    10c1c:       01052023                sw      a6,0(a0)
>    10c20:       00e82223                sw      a4,4(a6)
>    10c24:       0006a783                lw      a5,0(a3)
> ```
> 
> Mixture of `sw` and `lw`.

In this case, the memory address is not contiguous, what we can do just pack these instructions, but we cannot save any operation, such as checking misaligned.

qwe661234 replied:
> In addition, consider the following scenario:
> 
> ```
>    10a84:       01c12083                lw      ra,28(sp)
>    10a88:       07f47513                andi    a0,s0,127
>    10a8c:       01812403                lw      s0,24(sp)
>    10a90:       01412483                lw      s1,20(sp)
>    10a94:       01012903                lw      s2,16(sp)
>    10a98:       00c12983                lw      s3,12(sp)
> ```
> 
> It can be regarded as 5 `lw`. Roughly speaking, if [peephole optimization](https://en.wikipedia.org/wiki/Peephole_optimization) can be applied, we shall benefit from further optimizations.

In this case, we can pack the last four instruction lw. if we want to handle this case by packing 5 lw, we need to reorder the instruction. For example, swap the first and the second instruction.

qwe661234 replied:
> Can you handle the following case? (disassembly from CoreMark
> 
> ```
>    10248:       03012603                lw      a2,48(sp)
>    1024c:       01c11583                lh      a1,28(sp)
>    10250:       03412503                lw      a0,52(sp)
> ```

Ditto, if we want to handle this case, we need some strategies to reorder the instructions.

jserv replied:
In this pull request, let's concentrate on preliminary support of macro operation fusion. You shall add some comments for further efforts such as instruction reordering.

jserv commented on src/emulate.c:
```diff
@@ -1356,6 +1410,77 @@ static void block_translate(riscv_t *rv, block_t *block)
     block->ir[block->n_insn - 1].tailcall = true;
 }
 
+#define PACK_MEM_OPS(RW)                                                   \
```
The verb "pack" was not precise. Instead, use "combine," which conveys the notion of merging multiple instructions into a single one.

jserv commented on src/emulate.c:
```diff
@@ -1356,6 +1404,82 @@ static void block_translate(riscv_t *rv, block_t *block)
+            COMBINE_MEM_OPS(0);
+        case rv_insn_lw:
+            COMBINE_MEM_OPS(1);
+            /* FIXME: lui + addi*/
```
Should be `/* FIXME: lui + addi */`

jserv reviewed (changes requested):
Add some FIXME/TODO comments which address more macro operation fusion we can pay attention to.

jserv commented on src/emulate.c:
```diff
@@ -1356,6 +1404,82 @@ static void block_translate(riscv_t *rv, block_t *block)
+            next_ir->impl = dispatch_table[next_ir->opcode];               \
+        }                                                                  \
+    }                                                                      \
+    break;
```
Remove `break` here.

jserv commented on src/emulate.c:
```diff
@@ -1356,6 +1404,82 @@ static void block_translate(riscv_t *rv, block_t *block)
+         * are contiguous, combine these instructions.
+         */
+        case rv_insn_sw:
+            COMBINE_MEM_OPS(0);
```
Add `break` statement at the end of `COMBINE_MEM_OPS()`.

jserv reviewed (changes requested):
Check CI failure.

jserv reviewed (changes requested):
Check CI failure.

qwe661234 committed:
Introduce preliminary macro operation fusion

Through our observations, we have identified certain patterns in instruction
sequences. By converting these specific RISC-V instruction patterns into
faster and equivalent code, we can significantly improve execution efficiency.

In our current analysis, we focus on a commonly used benchmark and have
found the following frequently occurring instruction patterns: auipc + addi,
auipc + add, multiple sw, and multiple lw.

|  Metric  |     commit fba5802       |    macro fuse operation   |Speedup|
|----------+--------------------------+---------------------------+-------|
| CoreMark | 1351.065 (Iterations/Sec)|  1352.843 (Iterations/Sec)|+0.13% |
| dhrystone|       1073 DMIPS         |        1146 DMIPS         | +6.8% |
| nqueens  |       8295 msec          |        7824 msec          | +6.0% |

qwe661234 commented:
> Check CI failure.

In debug mode, the `rv_step` only emulates one instruction per step, specifically, it executes only the first instruction in a basic block then translate next basic block in PC + 4. If we apply macro fusion operations in debug mode, errors can occur. For instance, fusing auipc and addi and executing them together. However, the subsequent instruction is not a nop because the emulator only emulates the first instruction in a basic block. Consequently, the following instruction remains addi, resulting in an error because the result become auipc + addi + addi. 

Therefore, we cannot do fuse operation in debug mode.
